runtime/jcl/common/jclcinit.c

```diff
@@ -66,127 +66,128 @@ jint computeFullVersionString(J9JavaVM* vm)
 		if ((J2SE_VERSION(vm) & J2SE_RELEASE_MASK) == J2SE_19) {
-			strcat(fullversion, "9 ");
-			strcat(vminfo, "9 ");
+			j2se_version_info = "9";
 		} else {
-			strcat(fullversion, "9.? ");
-			strcat(vminfo, "9.? ");
+			j2se_version_info = "9.?";
 		}
 		break;
 	default:
-		strcat(fullversion, "?.?.? ");
-		strcat(vminfo, "?.?.? ");
+		j2se_version_info = "?.?.?";
 	}
 
-	strcat(fullversion, "IBM J9 ");
-	strcat(fullversion, EsVersionString " ");
 	osname = j9sysinfo_get_OS_type();
-	if (NULL != osname) {
-		strcat(fullversion, osname);
-		strcat(fullversion, " ");
-		strcat(vminfo, osname);
-		strcat(vminfo, " ");
-	}
 	osarch = j9sysinfo_get_CPU_architecture();
-	strcat(fullversion, osarch);
-	strcat(vminfo, osarch);
 
 #ifdef J9VM_ENV_DATA64
-	strcat(fullversion, "-64 ");
-	strcat(vminfo, "-64 ");
-#ifdef J9VM_GC_COMPRESSED_POINTERS
-	strcat(fullversion, "Compressed References ");
-	strcat(vminfo, "Compressed References ");
-#endif
+	#ifdef J9VM_GC_COMPRESSED_POINTERS
+		#define MEM_INFO "-64 Compressed References "
+	#else
+		#define MEM_INFO "-64 "
+	#endif
 #else
-#if defined(J9ZOS390) || defined(S390)
-	strcat(fullversion, "-31 ");
-	strcat(vminfo, "-31 ");
-#else
-	strcat(fullversion, "-32 ");
-	strcat(vminfo, "-32 ");
-#endif
+	#if defined(J9ZOS390) || defined(S390)
+		#define MEM_INFO "-31 "
+	#else
+		#define MEM_INFO "-32 "
+	#endif
 #endif
 
-	strcat(fullversion, EsBuildVersionString);
-	strcat(vminfo, EsBuildVersionString);
-
-#if defined(J9VM_INTERP_NATIVE_SUPPORT)
-	jitConfig = vm->jitConfig;
-	if (NULL != jitConfig) {
-		if (jitConfig->runtimeFlags & J9JIT_JIT_ATTACHED) {
-			jitEnabled = 1;
-		}
-		if (jitConfig->runtimeFlags & J9JIT_AOT_ATTACHED) {
-			aotEnabled = 1;
-		}
-	}
-	strcat(fullversion, " (JIT ");
-	strcat(fullversion, jitEnabled ? "en" : "dis");
-	strcat(vminfo, " (JIT ");
-	strcat(vminfo, jitEnabled ? "en" : "dis");
-	strcat(fullversion, "abled, AOT ");
-	strcat(fullversion, aotEnabled ? "en" : "dis");
-	strcat(vminfo, "abled, AOT ");
-	strcat(vminfo, aotEnabled ? "en" : "dis");
-	strcat(fullversion, "abled)\nOpenJ9   - ");
-	strcat(vminfo, "abled)\nOpenJ9   - ");
-#endif /* J9VM_INTERP_NATIVE_SUPPORT */
-
-	vmVersion = J9VM_VERSION_STRING;
-	strcat(fullversion, vmVersion);
-	strcat(vminfo, vmVersion);
-
 #if defined(J9VM_GC_MODRON_GC)
-	gcVersion = OMR_VERSION_STRING;
-	strcat(fullversion, "\nOMR      - ");
-	strcat(fullversion, gcVersion);
-	strcat(vminfo, "\nOMR      - ");
-	strcat(vminfo, gcVersion);
+	#define OMR_INFO "\nOMR      - " OMR_VERSION_STRING
+#else
+	#define OMR_INFO ""
 #endif /* J9VM_GC_MODRON_GC */
 
 #if defined(VENDOR_SHORT_NAME) && defined(VENDOR_SHA)
-	strcat(fullversion, "\n" VENDOR_SHORT_NAME "      - " VENDOR_SHA);
-	strcat(vminfo, "\n" VENDOR_SHORT_NAME "      - " VENDOR_SHA);
+	#define VENDOR_INFO "\n" VENDOR_SHORT_NAME "      - " VENDOR_SHA
+#else
+	#define VENDOR_INFO ""
 #endif /* VENDOR_SHORT_NAME && VENDOR_SHA */
 
+	if (bufferSize < j9str_printf(PORTLIB, NULL, 0,
```

Comment:
Remove this test and check the that return values from the two calls below are strictly less than `bufferSize` (to account for the trailing NUL).

Reply:
this will not work since j9str_printf will return `bufferSize - 1` when the string overflow

Reply:
You'll notice I said 'strictly less than bufferSize', i.e.
```
if (!(bufferSize > j9str_printf(...
```
or equivalently
```
if (bufferSize <= j9str_printf(...
```

Reply:
For the record, you were right: j9str_printf will never return a value larger than the limit it is given (when given a buffer). The test needs to be
```
if ((bufferSize - 1)<= j9str_printf(...
```
To be sure that there was room for the whole string and the trailing NUL.
Please make sure to include a comment pointing this out.